src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -127,6 +128,216 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
+  auto victim_wid = victim_addr.worker_id;
+
+  RAY_UNUSED(victim_client.StealWork(
+      std::move(request),
+      [this, scheduling_key, victim_wid, victim_addr, thief_addr, was_error,
+       assigned_resources](Status status, const rpc::StealWorkReply &reply) {
+        absl::MutexLock lock(&mu_);
+
+        RAY_CHECK(worker_to_lease_entry_.find(thief_addr) !=
+                  worker_to_lease_entry_.end());
+
+        auto &thief_entry = worker_to_lease_entry_[thief_addr];
+        RAY_CHECK(thief_entry.lease_client);
+        RAY_LOG(DEBUG) << "thief_entry.tasks_in_flight (thief): "
+                       << thief_entry.tasks_in_flight;
+        RAY_LOG(DEBUG) << "thief_entry.stealable_tasks.size() (thief): "
+                       << thief_entry.stealable_tasks.size();
+
+        size_t tasks_in_flight_before_stealing = thief_entry.tasks_in_flight;
+
+        ssize_t number_of_tasks_stolen = reply.number_of_tasks_stolen();
+        RAY_CHECK(number_of_tasks_stolen == reply.tasks_stolen_size());
+
+        RAY_LOG(DEBUG) << "We stole " << number_of_tasks_stolen << " tasks "
+                       << "from worker: " << victim_wid;
+
+        for (ssize_t i = 0; i < reply.tasks_stolen_size(); i++) {
+          const TaskSpecification task_spec(reply.tasks_stolen(i));
+          RAY_LOG(DEBUG) << "Thief " << thief_addr.worker_id << " Stole task "
+                         << task_spec.TaskId() << "from worker: " << victim_wid;
+
+          const SchedulingKey scheduling_key_check(
+              task_spec.GetSchedulingClass(), task_spec.GetDependencyIds(),
+              task_spec.IsActorCreationTask() ? task_spec.ActorCreationId()
+                                              : ActorID::Nil());
+          RAY_CHECK(scheduling_key_check == scheduling_key);
+
+          executing_tasks_.erase(task_spec.TaskId());
+
+          RAY_CHECK(worker_to_lease_entry_.find(victim_addr) !=
+                    worker_to_lease_entry_.end());
+          auto &victim_entry = worker_to_lease_entry_[victim_addr];
+          RAY_CHECK(victim_entry.lease_client);
+          victim_entry.stealable_tasks.erase(task_spec.TaskId());
+
+          auto &client = *client_cache_->GetOrConnect(thief_addr.ToProto());
+
+          RAY_CHECK(thief_entry.lease_client);
+          RAY_CHECK(thief_entry.tasks_in_flight ==
+                    tasks_in_flight_before_stealing + (size_t)i);
+
+          thief_entry.tasks_in_flight++;  // Increment the number of tasks in flight to
+                                          // the worker
+
+          auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+          scheduling_key_entry.total_tasks_in_flight++;
+
+          auto res = thief_entry.stealable_tasks.emplace(task_spec.TaskId());
+          RAY_CHECK(res.second);
+          executing_tasks_.emplace(task_spec.TaskId(), thief_addr);
+          PushNormalTask(thief_addr, client, scheduling_key, task_spec,
```

Comment:
How about reusing some of the existing code, to cut down on the additional logic? You could push the stolen tasks back onto the main queue, then call `OnWorkerIdle` for the thief worker, right? That would also help to make sure that we never exceed the maximum tasks in flight.